xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -248,10 +295,137 @@ private static ManagedChannel initLbChannel(
+    final String oldEdsServiceName;
+    @Nullable
+    final EndpointWatcher oldEndpointWatcher;
+
+    ClusterEndpointsBalancerProvider(String edsServiceName) {
+      this.edsServiceName = edsServiceName;
+      if (xdsConfig != null) {
+        oldEdsServiceName = xdsConfig.edsServiceName;
+      } else {
+        oldEdsServiceName = null;
+      }
+      oldEndpointWatcher = endpointWatcher;
+    }
+
+    @Override
+    public boolean isAvailable() {
+      return true;
+    }
+
+    @Override
+    public int getPriority() {
+      return 5;
+    }
+
+    // A synthetic policy name identified by edsServiceName in XdsConfig.
+    @Override
+    public String getPolicyName() {
+      return "xds_policy__edsServiceName_" + edsServiceName;
+    }
+
+    @Override
+    public LoadBalancer newLoadBalancer(Helper helper) {
+      return new ClusterEndpointsBalancer(helper);
+    }
+
+    /**
+     * Load-balances endpoints for a given cluster.
+     */
+    final class ClusterEndpointsBalancer extends LoadBalancer {
+      final Helper helper;
+
+      // All fields become non-null once handleResolvedAddresses() successfully.
+      // All fields are assigned at most once.
+      @Nullable
+      LocalityStore localityStore;
+      @Nullable
+      LoadReportClient lrsClient;
+      @Nullable
+      EndpointWatcherImpl endpointWatcher;
+
+      ClusterEndpointsBalancer(Helper helper) {
+        this.helper = helper;
+      }
+
+      @Override
+      public void handleNameResolutionError(Status error) {
+        // Go into TRANSIENT_FAILURE if we have not yet received any endpoint update. Otherwise,
+        // we keep running with the data we had previously.
+        if (endpointWatcher == null || !endpointWatcher.firstEndpointUpdateReceived) {
+          helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+        }
+      }
+
+      @Override
+      public boolean canHandleEmptyAddressListFromNameResolution() {
+        return true;
+      }
+
+      @Override
+      public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+        XdsConfig xdsConfig = (XdsConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+
+        if (endpointWatcher != null) {
+          // TODO(zddapeng): Handle child policy changed if any.
+          return;
+        }
+
+        LoadStatsStore loadStatsStore = new LoadStatsStoreImpl();
+        localityStore = localityStoreFactory.newLocalityStore(helper, lbRegistry, loadStatsStore);
+        LoadReportCallback lrsCallback =
+            new LoadReportCallback() {
+              @Override
+              public void onReportResponse(long reportIntervalNano) {
+                localityStore.updateOobMetricsReportInterval(reportIntervalNano);
+              }
+            };
+
+        // TODO(zdapeng): Use XdsClient to do Lrs directly.
+        // For now create an LRS Client.
+        if (xdsConfig.balancerName != null) {
+          ManagedChannel channel = initLbChannel(
```

Comment:
You shouldn't be creating a new channel for LRS through. It should share the same channel created at line 189. This is the legacy case using balancer name from service config. I found it hard to maintain two separate code paths with one using `XdsClient` interface and the other not, while the one not using `XdsClient` has two places sharing the same channel. It's messy, may be eliminate that in this PR is easier? We probably accept the tests to be broken temporarily (in worst case).

Reply:
Cached channel in the balancer.